### PR TITLE
spice up the hosted game info

### PR DESCRIPTION
**Pull request by GrotheFAF**
add teams + observer info (behind N/M Player)
color the player+team if enough player and/or team match
add sim mods hint
add short highlight on game update
add host time (as not from server [yet] - hosted games on client start get that time)
niced tooltip - signal installed sim-mods in green

### CODE REVIEW

File: src/games/gamemodelitem.py [hunk 1]
```diff
@@ -17,9 +17,26 @@ def __init__(self, game, me):
         self._me = me
         self._me.relationsUpdated.connect(self._check_host_relation_changed)
 
+        # Update highlight
+        self._highlight_timer = QTimer()
+        self._highlight_timer.setSingleShot(True)
+        self._highlight_timer.setInterval(250)
+        self._highlight_timer.timeout.connect(self._game_updated)
```
Comment (Wesmania) [inline src/games/gamemodelitem.py]: I'd connect it to a new method in case we ever want to do something more at actual game update.

Comment (GrotheFAF) [reply src/games/gamemodelitem.py]: yes, splitting it is better.

File: src/games/gameitem.py [hunk 2]
```diff
@@ -268,13 +341,17 @@ def _player_table_row(self, player, align):
 
     def _country_icon_fmt(self, player):
         icon_path_fmt = os.path.join("chat", "countries", "{}.png")
-        icon_path = icon_path_fmt.format(player.country.lower())
+        if player.country is None or player.country == '':
```
Comment (ChrisKitching) [inline src/games/gameitem.py:344]: Either the system is quite insanely designed, or at least one of these possibilities shouldn't exist?

I would suggest just enforcing that `country is None` when unknown, since then this if-block can be replaced with `country = (player.country or '__').lower()` or such.

File: src/games/gameitem.py [hunk 3]
```diff
@@ -268,13 +341,17 @@ def _player_table_row(self, player, align):
...
+        else:
+            country = player.country.lower()
+        icon_path = icon_path_fmt.format(country)
         icon_abs_path = os.path.join(util.COMMON_DIR, icon_path)
         return "<img src='{}'>".format(icon_abs_path)
 
     def _player_fmt(self, player):
         if player == self._me.player:
-            pformat = "<b><i>{}</b></i>"
+            pformat = "<b><i>{}</i></b>"
```
Comment (ChrisKitching) [inline src/games/gameitem.py:354]: This seems like the sort of obvious fix that could be its own commit, and merged independently of this giant monolithic commit.

Giant monolithic commits are extremely hard to review. This may be why so many of the long-open PRs on this repository are the ones that are like that :D

File: src/games/gameitem.py [hunk 4]
```diff
@@ -286,30 +363,31 @@ def _observers_tooltip(self, observers):
...
-        return "Observers: " + ", ".join(observer_strings)
+        return obs_table_start + ",</td>".join(observer_strings) + obs_table_end
 
     def _mods_tooltip(self, mods):
         if not mods:
             return ""
-        return "<br/>With: " + "<br/>".join(mods.values())
+        mods_start = "<br/>With Sim Mods:"
+        mods_inst = "<br/><font color='green'><i>Installed:</i><br/>" + \
+                    "<br/>".join(mods[0]) + "</font>" if mods[0] else ""
```
Comment (ChrisKitching) [inline src/games/gameitem.py:381]: Instead of using `mods[0]` for installed mods and `mods[1]` for not-installed mods, it would be much less insane to use a named dictionary field. So people don't have to stare at this for 30 seconds before they figure out what the hell is going on, as I just did.

You could go one better and make it an actual object, and enforce that it's never `None`. This would allow you to avoid the checks (since `join()` with empty-list yields empty-string).

File: src/games/gamemodelitem.py [hunk 5]
```diff
@@ -15,13 +16,34 @@ def __init__(self, game, me, preview_dler):
...
         self.game = game
         self.game.gameUpdated.connect(self._game_updated)
+        self.hosted_at = None  # put here as long it's not coming from server
         self._me = me
         self._me.relationsUpdated.connect(self._check_host_relation_changed)
         self._preview_dler = preview_dler
         self._preview_dl_request = PreviewDownloadRequest()
         self._preview_dl_request.done.connect(self._at_preview_downloaded)
 
+        # Update highlight
```
Comment (ChrisKitching) [inline src/games/gamemodelitem.py:26]: The complete lack of comments explaining what this "highlight" logic is trying to achieve is unhelpful.

File: src/games/gameitem.py [hunk 6]
```diff
@@ -209,11 +261,27 @@ def _game_observers(self, game):
         return [game.to_player(name) for name in game.observers
                 if game.is_connected(name)]
 
+    def _game_sim_mods(self, game):
+        if not game.sim_mods:
```
Comment (ChrisKitching) [inline src/games/gameitem.py:265]: Another example of oddly defensive programming.

What's the contract of the `game` object? It would appear sensible to make it always have the `sim_mods` field, and simply set it to empty-list when there are no mods. Having to check for existence of fields all the time is a really good bug factory.

File: src/games/gameitem.py [hunk 7]
```diff
@@ -209,11 +261,27 @@ def _game_observers(self, game):
...
+            return None
+        inst = modvault.getInstalledMods()
+        uids = [mod.uid for mod in inst]
+        mods = ([], [])
```
Comment (ChrisKitching) [inline src/games/gameitem.py:269]: Oh god why: we have these things called "objects". :D

File: src/games/gameitem.py [hunk 8]
```diff
@@ -209,11 +261,27 @@ def _game_observers(self, game):
         return [game.to_player(name) for name in game.observers
                 if game.is_connected(name)]
 
+    def _game_sim_mods(self, game):
```
Comment (ChrisKitching) [inline src/games/gameitem.py:264]: Doccomment? What is this function for?

File: src/games/gameitem.py [hunk 9]
```diff
@@ -209,11 +261,27 @@ def _game_observers(self, game):
...
+        inst = modvault.getInstalledMods()
+        uids = [mod.uid for mod in inst]
+        mods = ([], [])
+        for uid in game.sim_mods:
+            if uid in uids:
+                mods[0].append(game.sim_mods[uid])
+            else:
+                mods[1].append(game.sim_mods[uid])
+            mods[0].sort()
+            mods[1].sort()
```
Comment (ChrisKitching) [inline src/games/gameitem.py:276]: Why are you sorting both lists after any insertion into either list? Just sort the lists at the end, surely?!

File: src/games/gameitem.py [hunk 10]
```diff
@@ -209,11 +261,27 @@ def _game_observers(self, game):
         return [game.to_player(name) for name in game.observers
                 if game.is_connected(name)]
 
+    def _game_sim_mods(self, game):
+        if not game.sim_mods:
+            return None
+        inst = modvault.getInstalledMods()
+        uids = [mod.uid for mod in inst]
+        mods = ([], [])
+        for uid in game.sim_mods:
```
Comment (ChrisKitching) [inline src/games/gameitem.py:270]: Loop can be replaced with two list comprehensions.

File: src/games/gamemodelitem.py [hunk 11]
```diff
@@ -15,13 +16,34 @@ def __init__(self, game, me, preview_dler):
 
         self.game = game
         self.game.gameUpdated.connect(self._game_updated)
+        self.hosted_at = None  # put here as long it's not coming from server
```
Comment (ChrisKitching) [inline src/games/gamemodelitem.py:19]: Unclear why you're proposing to add code for a feature that definitely won't work. Maybe add that part after filing the corresponding server PR?

**Comment by Wesmania** [comment]
Replaced with jinja.
